### ISSUE TITLE
non_max_suppression on model results

### ISSUE THREAD

**Issue by erex16** [ultralytics/ultralytics#2556]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I am aware that non_max_suppression is already called within the code, however I would like to run nms again after the initial model predictions. I referred to the code in the postprocess function of DetectionPredictor to see how to call the function and do subsequent postprocessing. However, the results aren't making sense. Here is how I am calling the function as well as the return. 

pred = self.model(image, augment=False)
new_preds = ops.non_max_suppression(pred[0].boxes.data,
                                                                conf_thres, 
                                                                iou_thres, 
                                                                classes=[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18])

for i, pred in enumerate(new_preds):
        if not isinstance(image, torch.Tensor):
                pred[:, :4] = ops.scale_boxes(torch.Size([736, 736]), pred[:, :4], image.shape) # Model trained on imgsz=736
                print(pred.cpu().detach().numpy())

This is an example of the output I get from this :
[[     1412.3      379.51        1920        1080           8           1]]
[[     412.39      409.17        1920        1080           7           1]]

I believe the formation should be [x1, y1, x2, y2, class, conf] but I don't understand why all my x2/y2 coordinates end up as the image width/height. The bounding boxes I get from the initial self.model call are correct. So am I calling non_max_suppression incorrectly? Post processing incorrectly? I was unable to find any further documentation on how to do this so please let me know if there is somewhere else to look.

**Comment by glenn-jocher** [comment]
@erex16 hi there!

It sounds like you're trying to call non-maximum suppression (NMS) again after the initial model predictions, but the results aren't making sense. Based on the code you provided, it looks like you are calling `ops.non_max_suppression` with the correct arguments, but the formation of your output doesn't look quite right.

To answer your question, I believe the formation should be [x1, y1, x2, y2, conf, class], so you might want to swap the last two elements of the array? It seems unusual that all your x2/y2 coordinates end up as the image width/height, so you might want to double-check your post-processing code to make sure you're scaling the boxes correctly.

If you're still having issues, feel free to provide more information on what you are trying to achieve and how you would like your output formatted, and we can try to provide more guidance.

Hope this helps!

**Comment by M-Amrollahi** [comment]
> Based on the code you provided, it looks like you are calling `ops.non_max_suppression` with the correct arguments, but the formation of your output doesn't look quite right.

I see a lot of bboxes overlapped although I have set ```nms: True``` with ```iou: .7```. Should I call the ```non_max_suppression``` after ```model.predict(...)```? How should I pass the predictions to this function?

```python
res = model.predict(x)
```

**Comment by glenn-jocher** [comment]
@M-Amrollahi if you have already set `nms: True` with `iou: .7`, non-maximum suppression (NMS) should be automatically applied during the model's prediction phase. Therefore, you normally shouldn't need to call `non_max_suppression` again after `model.predict(...)`. 

However, if you are seeing a lot of overlapping bounding boxes, then it could be due to several reasons. For instance, it may be that the confidence threshold set for the predictions is too low, which could lead to many weak predictions that overlay stronger ones. Another possibility could be that the classes of the overlapping bounding boxes are different, since NMS operates on a per-class basis.

If you're still not getting your desired results and you'd like to try manually applying NMS again, you could in principle do so. Your model's predictions (`res` in your example code) will typically include the bounding box coordinates and their associated confidence scores and class predictions. These are the inputs you'd need for `non_max_suppression`. Please refer to the specific implementation or documentation of `non_max_suppression` function in your codebase if you need more specific information on how to format these inputs.

**Comment by M-Amrollahi** [comment]
> [glenn-jocher](/glenn-jocher)

Does the ```non_max_suppression``` apply on predicted results automatically? As I saw the sources, the ```non_max_suppression``` will be called on ```postprocess``` but there is no one call ```postprocess```. Should we call it manually? If yes, I do not know how will be the input shape and type? I believe that it should be like :
```
[[x1 y1 x2 y2 conf cls]]
```
I do not understand the [description](https://docs.ultralytics.com/reference/utils/ops/#ultralytics.utils.ops.non_max_suppression) on the reference:
```
A tensor of shape (batch_size, num_classes + 4 + num_masks, num_boxes) containing the predicted boxes, classes, and masks.
The tensor should be in the format output by a model, such as YOLO.
```

Could you tell me where I am wrong here in this part of code?
```python
results = model.predict(val_image)

hits = torch.cat((results[0].boxes.xyxy,
                      results[0].boxes.conf.reshape(-1,1),
                      results[0].boxes.cls.reshape(-1,1)),dim=1)

hits = hits.unsqueeze(dim=0)
    
nms_hits = non_max_suppression(hits, conf_thres = CONFID, iou_thres = IOU_THRESH, nc=5)
```

**Comment by glenn-jocher** [comment]
@M-Amrollahi the `non_max_suppression` function is typically called as part of the post-processing step in a YOLO detector. During this step, each predicted bounding box's confidence score is compared against a confidence threshold, and if it's too low, the bounding box is discarded. Next, the non-max suppression is conducted to remove duplicates and retain the most probable bounding box when multiple bounding boxes significantly overlap and target the same object.

As for the location of `postprocess` call, it normally would be called by a higher-level function or method that encompasses the end-to-end process of generating a prediction from an input image. 

Regarding the input to the `non_max_suppression`, it indeed should be in the format of `[x1, y1, x2, y2, conf, cls]` for each bounding box. When it is stated that the tensor should be in the format output by a model, such as YOLO, it means that these elements must come directly from the model’s output predictions. 

The code you've posted is essentially concatenating these elements correctly and forming `hits` tensor as per the expectations. However, it's unclear where `CONFID`, `IOU_THRESH`, and `nc=5` are coming from or what they represent.

In case you notice a lot of overlapping bounding boxes, please make sure that the confidence threshold and IOU threshold are correctly set. Additional overlapping may arise if the classes of the bounding boxes are different as NMS is class-specific. Thus, adjust these hyperparameters according to your dataset and requirements.

To sum up, you generally don't need to call `non_max_suppression` manually if it's already part of the YOLO model's pipeline/process. The import convention and setup you have should suffice unless the running model’s specifications demand otherwise.

**Comment by M-Amrollahi** [comment]
I have some problems with using ```non_max_suppression``` as I do not understand correctly the input and the output.
I have tried [nms in pytorch](https://pytorch.org/vision/main/generated/torchvision.ops.nms.html) which is more clear to me.
Thanks for your helps🙏

**Comment by glenn-jocher** [comment]
Hello @M-Amrollahi,

I understand that you're having some difficulties with the `non_max_suppression` function in YOLOv8. This function is part of the post-processing step, and it is generally called automatically when you perform model predictions, so normally, you wouldn't need to call it again manually.

The function is used to remove any redundant or overlapping bounding boxes from the model's predictions. It operates on the confidence scores, coordinates, and classes of each predicted bounding box.

The input to `non_max_suppression` usually should be in the format of `[x1, y1, x2, y2, conf, cls]` for each bounding box directly from the model's output predictions. Here, `x1`, `y1`, `x2`, `y2` refer to the coordinates of your bounding box, `conf` is the confidence score of the bounding box, and `cls` is the class prediction.

If you find a lot of overlapping bounding boxes, do verify if the confidence threshold (conf) and IOU threshold are appropriately set. You might also want to check the classes of the bounding boxes since NMS operates on a per-class basis.

I hope this helps clarify the usage of `non_max_suppression`. But if you're still encountering issues, please feel free to provide more specifics, and we'd be happy to assist you further.

**Comment by M-Amrollahi** [comment]
I doubt that it is applied automatically as I said I find nowhere that
calls this function. Also here in this blog, he used this function manually.

https://kean-chan.medium.com/performing-inference-with-yolov8-and-retraining-with-custom-dataset-84d75c304531

I think that is true and thid function should not be part of prediction. It
must be separate.

Thanks
